File: backend/app/schemas/availability.py
import uuid
from datetime import date, time
from pydantic import BaseModel, field_validator, model_validator


class AvailabilityRuleCreate(BaseModel):
    day_of_week: int  # 0=Mon, 6=Sun
    start_time: time
    end_time: time
    slot_duration_minutes: int = 30
    buffer_minutes: int = 0
    max_bookings_per_day: int | None = None
    is_active: bool = True
    timezone: str = "America/Santiago"

    @field_validator("day_of_week")
    @classmethod
    def validate_dow(cls, v: int) -> int:
        if not 0 <= v <= 6:
            raise ValueError("day_of_week must be 0–6 (Monday=0, Sunday=6)")
        return v

    @model_validator(mode="after")
    def validate_time_range(self) -> "AvailabilityRuleCreate":
        if self.end_time <= self.start_time:
            raise ValueError("end_time must be after start_time")
        return self


class AvailabilityRuleResponse(BaseModel):
    id: uuid.UUID
    day_of_week: int
    start_time: time
    end_time: time
    slot_duration_minutes: int
    buffer_minutes: int
    max_bookings_per_day: int | None
    is_active: bool
    timezone: str
    model_config = {"from_attributes": True}


class WeeklyScheduleRequest(BaseModel):
    rules: list[AvailabilityRuleCreate]


class WeeklyScheduleResponse(BaseModel):
    rules: list[AvailabilityRuleResponse]


class OverrideCreateRequest(BaseModel):
    override_date: date
    is_closed: bool = False
    custom_start_time: time | None = None
    custom_end_time: time | None = None
    reason: str | None = None


class OverrideResponse(BaseModel):
    id: uuid.UUID
    override_date: date
    is_closed: bool
    custom_start_time: time | None
    custom_end_time: time | None
    reason: str | None
    model_config = {"from_attributes": True}


class SlotItem(BaseModel):
    start: str  # ISO8601 with timezone offset
    end: str
    available: bool


class SlotListResponse(BaseModel):
    date: str
    slots: list[SlotItem]

File: frontend/lib/api/availability.ts
// frontend/lib/api/availability.ts
import api from '@/lib/api'

export interface AvailabilityRule {
  id: string
  day_of_week: number  // 0=Mon, 6=Sun
  start_time: string   // "HH:MM:SS"
  end_time: string
  slot_duration_minutes: number
  buffer_minutes: number
  max_bookings_per_day: number | null
  is_active: boolean
  timezone: string
}

export interface AvailabilityRuleInput {
  day_of_week: number
  start_time: string
  end_time: string
  slot_duration_minutes: number
  buffer_minutes: number
  max_bookings_per_day?: number | null
  is_active: boolean
  timezone: string
}

export interface AvailabilityOverride {
  id: string
  override_date: string
  is_closed: boolean
  custom_start_time: string | null
  custom_end_time: string | null
  reason: string | null
}

export interface AvailabilityOverrideInput {
  override_date: string
  is_closed: boolean
  custom_start_time?: string | null
  custom_end_time?: string | null
  reason?: string | null
}

export interface Slot {
  start: string
  end: string
  available: boolean
}

export interface SlotListResponse {
  date: string
  slots: Slot[]
}

export const availabilityApi = {
  getRules: async (): Promise<{ rules: AvailabilityRule[] }> => {
    const { data } = await api.get('/api/v1/availability/rules')
    return data
  },

  upsertRules: async (rules: AvailabilityRuleInput[]): Promise<{ rules: AvailabilityRule[] }> => {
    const { data } = await api.put('/api/v1/availability/rules', { rules })
    return data
  },

  getSlots: async (date: string, serviceId: string): Promise<SlotListResponse> => {
    const { data } = await api.get('/api/v1/availability/slots', {
      params: { date, service_id: serviceId },
    })
    return data
  },

  getOverrides: async (fromDate: string, toDate: string): Promise<AvailabilityOverride[]> => {
    const { data } = await api.get('/api/v1/availability/overrides', {
      params: { from_date: fromDate, to_date: toDate },
    })
    return data
  },

  createOverride: async (input: AvailabilityOverrideInput): Promise<AvailabilityOverride> => {
    const { data } = await api.post('/api/v1/availability/overrides', input)
    return data
  },

  deleteOverride: async (overrideId: string): Promise<void> => {
    await api.delete(`/api/v1/availability/overrides/${overrideId}`)
  },
}
